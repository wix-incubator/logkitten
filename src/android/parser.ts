import { Parser } from '../types';
import { LineBuffer } from '../utils';
import { AndroidEntry } from './types';
import { areEntriesComingFromTheSameSource, PriorityUtils } from './utils';

export class AndroidParser {
  private _buffer = new LineBuffer();
  // In fact, this is: [] | [AndroidEntry]
  private _pendingEntry: AndroidEntry[] = [];

  private static readonly logcatRegex: RegExp =
    //   epoch        uid     pid     tid     level   tag     message
    /^\s*(\d+\.\d+)\s+(\S+)\s+(\d+)\s+(\d+)\s+(\w)\s+(.+?)\s*:\s*(.*)$/m;

  static create(): Parser<AndroidEntry> {
    const parser = new AndroidParser();
    return parser.parse.bind(parser);
  }

  parse(raw: string | Buffer, stderr: boolean): AndroidEntry[] {
    if (stderr) {
      return [];
    }

    const str = typeof raw === 'string' ? raw : raw.toString('utf-8');
    const lines = this._buffer.processChunk(str);
    return this._parseMessages(lines);
  }

  private _parseMessages(messages: string[]): AndroidEntry[] {
    const newEntries = messages.map(
      (rawMessage: string): AndroidEntry | null => {
        const match = rawMessage.match(AndroidParser.logcatRegex);
        if (match) {
          const [, epoch, uid, pid, tid, level, tag, message] = match;

          return {
            ts: Math.floor(parseFloat(epoch) * 1000),
            pid: Number.parseInt(pid, 10),
            tid: Number.parseInt(tid, 10),
            level: PriorityUtils.fromLetter(level),
            tag: tag.trim() || 'unknown',
            msg: message.trim(),
            uid,
          };
        }

        return null;
      }
    );

    const result = newEntries.reduce((acc, entry) => {
      if (!entry) {
        return acc;
      }

      const last = acc.at(-1);
      if (last && areEntriesComingFromTheSameSource(last, entry)) {
        last.msg += '\n' + entry.msg;
        return acc;
      }

      acc.push(entry);

      return acc;
    }, this._pendingEntry);

    this._pendingEntry = result.splice(-1);
    return result;
  }
}
