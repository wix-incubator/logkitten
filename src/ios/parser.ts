import { LineBuffer } from '../utils';
import { Parser } from '../types';
import type { IosEntry, RawEntryIOS } from './types';
import { PriorityUtils } from './utils';

const IGNORED_STDERR_PATTERN = /\bgetpwuid_r\b/;

export class IosParser {
  private _buffer = new LineBuffer();

  constructor() {
    this.parse = this.parse.bind(this);
  }

  static create(): Parser<IosEntry> {
    const parser = new IosParser();
    return parser.parse.bind(parser);
  }

  parse(raw: string | Buffer, stderr: boolean): IosEntry[] {
    const str = typeof raw === 'string' ? raw : raw.toString('utf-8');
    const lines = this._buffer.processChunk(str);
    return lines.flatMap((line) => this._parseLine(line, stderr));
  }

  private _parseLine(line: string, stderr: boolean): IosEntry[] {
    if (stderr) {
      return this._parseStderr(line);
    }

    let rawEntry: RawEntryIOS;
    try {
      rawEntry = JSON.parse(line);
    } catch {
      return [];
    }

    const entry: IosEntry = {
      msg: rawEntry.eventMessage || '',
      ts: Date.parse(rawEntry.timestamp),
      pid: rawEntry.processID,
      tid: rawEntry.threadID,
      level: PriorityUtils.fromName(rawEntry.messageType),
      // Platform specific fields
      category: rawEntry.category || '',
      formatString: rawEntry.formatString || '',
      processImagePath: rawEntry.processImagePath || '',
      subsystem: rawEntry.subsystem || '',
      userID: rawEntry.userID,
    };

    return [entry];
  }

  private _parseStderr(line: string): IosEntry[] {
    if (line.includes('No devices are booted.')) {
      throw new Error('No simulators are booted.');
    }

    if (IGNORED_STDERR_PATTERN.test(line)) {
      return [];
    }

    throw new Error(line);
  }
}
