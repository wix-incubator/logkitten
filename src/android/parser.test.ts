import { Chunker } from '../__tests__/Chunker';
import type { AndroidEntry } from './types';
import { AndroidParser } from './parser';

const DUMP = `
--------- beginning of main
                  0.785  1000  1321  1321 I SELinux : SELinux: Loaded service_contexts from:
--------- beginning of events
                  0.785  1000  1321  1321 I auditd  : SELinux: Loaded service_contexts from:
--------- beginning of radio
         1750850005.865160  1000   125   125 D snet_event_log: [subtag=121035042,-1,]
         1750850005.871666  lmkd   124   124 I lowmemorykiller: Using psi monitors for memory pressure detection
         1750850005.871771  lmkd   124   124 V lowmemorykiller: Process polling is supported
         1750850005.871996  lmkd   124   124 V lowmemorykiller: Process reaper initialized with 2 threads in the pool
         1750850005.872895  1000   125   125 D auditd  : avc=SELinux: Loaded service context from:
         1750850005.872898  1000   125   125 D auditd  : avc=		/system/etc/selinux/plat_service_contexts
         1750850005.872899  1000   125   125 D auditd  : avc=		/system_ext/etc/selinux/system_ext_service_contexts
         1750850005.872900  1000   125   125 D auditd  : avc=		/vendor/etc/selinux/vendor_service_contexts
         1750850005.874787  root   127   127 I qemu-props: successfully set property 'vendor.qemu.sf.fake_camera' to 'front'
         1750850005.874815  root   127   127 I qemu-props: successfully set property 'qemu.hw.mainkeys' to '0'
         1750850005.874833  root   127   127 I qemu-props: successfully set property 'qemu.sf.lcd_density' to '420'
         1750850005.874833  lmkd   124   124 W lowmemorykiller: This message won't make it to the snapshots because of buffering
`;

describe('AndroidParser chunked input', () => {
  it('parses entries correctly with chunked input', () => {
    const parser = new AndroidParser();
    // Use the first fixture as a sample
    const chunker = new Chunker(DUMP, 4);
    let entries: AndroidEntry[] = [];

    for (const chunk of chunker.getChunks()) {
      entries = entries.concat(parser.parse(chunk, false));
    }

    expect(entries).toMatchSnapshot();
  });

  it('ignores stderr', () => {
    const parser = new AndroidParser();
    const entries = parser.parse(DUMP, true);
    expect(entries).toEqual([]);
  });
});
