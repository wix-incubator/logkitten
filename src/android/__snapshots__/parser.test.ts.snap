// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AndroidParser chunked input parses entries correctly with chunked input 1`] = `
[
  {
    "level": 30,
    "msg": "SELinux: Loaded service_contexts from:",
    "pid": 1321,
    "tag": "SELinux",
    "tid": 1321,
    "ts": 785,
    "uid": "1000",
  },
  {
    "level": 30,
    "msg": "SELinux: Loaded service_contexts from:",
    "pid": 1321,
    "tag": "auditd",
    "tid": 1321,
    "ts": 785,
    "uid": "1000",
  },
  {
    "level": 20,
    "msg": "[subtag=121035042,-1,]",
    "pid": 125,
    "tag": "snet_event_log",
    "tid": 125,
    "ts": 1750850005865,
    "uid": "1000",
  },
  {
    "level": 30,
    "msg": "Using psi monitors for memory pressure detection",
    "pid": 124,
    "tag": "lowmemorykiller",
    "tid": 124,
    "ts": 1750850005871,
    "uid": "lmkd",
  },
  {
    "level": 10,
    "msg": "Process polling is supported
Process reaper initialized with 2 threads in the pool",
    "pid": 124,
    "tag": "lowmemorykiller",
    "tid": 124,
    "ts": 1750850005871,
    "uid": "lmkd",
  },
  {
    "level": 20,
    "msg": "avc=SELinux: Loaded service context from:
avc=		/system/etc/selinux/plat_service_contexts
avc=		/system_ext/etc/selinux/system_ext_service_contexts
avc=		/vendor/etc/selinux/vendor_service_contexts",
    "pid": 125,
    "tag": "auditd",
    "tid": 125,
    "ts": 1750850005872,
    "uid": "1000",
  },
  {
    "level": 30,
    "msg": "successfully set property 'vendor.qemu.sf.fake_camera' to 'front'
successfully set property 'qemu.hw.mainkeys' to '0'
successfully set property 'qemu.sf.lcd_density' to '420'",
    "pid": 127,
    "tag": "qemu-props",
    "tid": 127,
    "ts": 1750850005874,
    "uid": "root",
  },
]
`;
